plugins/Goals/Commands/CalculateConversionPages.php

```diff
@@ -0,0 +1,339 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\Goals\Commands;
+
+use Piwik\Common;
+use Piwik\Container\StaticContainer;
+use Piwik\Date;
+use Piwik\Db;
+use Piwik\Plugin\ConsoleCommand;
+use Piwik\Plugins\Goals\Model as GoalsModel;
+use Piwik\Plugins\SitesManager\Model as SitesModel;
+use Piwik\Site;
+use Piwik\Timer;
+use Piwik\Tracker\GoalManager;
+use Piwik\Updater;
+use Piwik\Updater\Migration\Factory as MigrationFactory;
+
+/**
+ * Command to calculate the pages viewed before conversions and populate the log_conversion.pages_before field
+ */
+class CalculateConversionPages extends ConsoleCommand
```

Comment:
The code in this class is currently fully untested, right?
Was wondering if we maybe could add a simple test for that. Like tracking a couple of visits & conversions. Afterwards unsetting the `pageviews_before` column in the database and then running the command to check if the column is filled up correctly again...

Reply:
I've added a new test `CalculateConversionPagesCommandTest`. It uses the same fixture as the `TrackGoalsPagesTest` which includes some of the more complicated conversion scenarios. Once the fixture is setup the test clears the `pageviews_before` column and then runs the command (both in date range mode and lastN) then compares the newly calculated `pageviews_before` values against the known values from the `TrackGoalsPagesTest`.

This test revealed an issue with the update query for visits which have multiple conversion of different goals, I've adjusted the query to calculate correctly for these conversions. The updated query is more complicated, so after optimizations I've retested it against a large dataset and it is updating 100,000 conversions with 4.5million linked visit actions in ~22 seconds which would still seem to be an acceptable level of performance.